src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp

```diff
@@ -50,47 +57,99 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
 
   // Loop through all possible split points, choosing the best one.  Also, force
   // a minimum leaf size of 1 (empty children don't make sense).
-  double bestFoundGain = bestGain;
+  double bestFoundGain = std::min(bestGain + minimumGainSplit, 0.0);
+  bool improved = false;
   const size_t minimum = std::max(minimumLeafSize, (size_t) 1);
-  for (size_t index = minimum; index < data.n_elem - (minimum - 1); ++index)
+
+  // We need to count the number of points for each class.
+  arma::Mat<size_t> classCounts;
+  arma::mat classWeightSums;
+  double totalWeight;
+  double totalLeftWeight = 0.0;
+  double totalRightWeight = 0.0;
+  if (UseWeights)
+  {
+    classWeightSums.zeros(numClasses, 2);
+    totalWeight = arma::accu(sortedWeights);
+    bestFoundGain *= totalWeight;
+  }
+  else
+  {
+    classCounts.zeros(numClasses, 2);
+    bestFoundGain *= data.n_elem;
+  }
+
+  // Initialize the counts.
+  // These points have to be on the left.
+  for (size_t i = 0; i < minimum - 1; ++i)
+  {
+    if (UseWeights)
+    {
+      classWeightSums(sortedLabels[i], 0) += sortedWeights[i];
```

Comment:
We could move this into the if/else case from above, to make this shorter.

Reply:
Yeah, agreed.  Done in 7b584bde.  I'm not sure it's actually shorter (didn't check too closely) but it's certainly easier to read.